### PR TITLE
Feature: Enable access controlled links

### PR DESCRIPTION
**Pull request by NinaHerrmann**
WIP!
This pull request aims to insert the function of a controlled link/access controlled link. This link ensures that after submission files can merely restrictedly changed and the access of the file is controlled with private links rather than public links. 
The Google drive and OneDrive repositories are examples where the functionality has been implemented.

Currently the functionalities are not sufficiently tested.

### CODE REVIEW

**Comment by NinaHerrmann** [comment]
To increase clarity, the essential tasks to be done before the core functionality is usable are listed:

- [x] **Pop-up window for authentication**
- [x] **File is moved as a share in the users account** 
- [x] **Settings-field for the name of the folder in owncloud.** The default value is _Moodlefiles_.
- [x] **Settings-field for the time a file should be shared.** The default value is 1 week.
- [x] **Handle moving a file when file is renamed in case file with same name already exists.** when file is moved the information about the share is requested which includes possible renaming.
- [x] **Redirection url is built with the file-id.** Webdav redirect enforces download.
- [x] **Redirect to download** Done with a header to the webdav-dowload-url.
- [x] **Display different file types** Redirecting with webdav has the major benefit that pictures are displayed embedded. 
- [x] **Handle multiple files** Not considered as a severe problem, since normal files do not offer that functionality.

**Comment by NinaHerrmann** [comment]
The todos for the next commits:
- [x] **Discuss how errors should be displayed** (e.g. When file can no longer be accessed) possible solutions: email admin, show message to user, contact course admin? (see .pdf files)
- [x] **Insert basic phpunit test for all functions and fix existing test.** 

- [ ] **Users can not differentiate between _normal_ files and _access_controlled links_.** This is disturbing for the user, since a file should merely be shared with the user and downloaded when he/she knowns about the procedure. This is partly handled by printing an information message, however, this is merely displayed on the next Moodle page loaded.
- [ ] **Handle login in case of images** Images are being displayed on the Moodle page. However, this suppresses the login button (and the corresponding pop-up) in case the user is not authenticated. Therefore, no content is displayed. The google repository has the same problem. I see three ways to solve this issue:

1. Firstly, the display type of the resource could be changed from `RESOURCELIB_DISPLAY_AUTO` (which transfers to `RESOURCELIB_DISPLAY_EMBED` look at `mod/resource/view.php`) to e.g. `RESOURCELIB_DISPLAY_DOWNLOAD` or `RESOURCELIB_DISPLAY_FRAME`. Therefore, the corresponding database entry needs to be updated in the function `reference_file_selected()`. It might be troublesome to get the id for updating the file.
2. Secondly, `send_file()` could be adjusted to behave differently for pictures. However, I am unsure to which extent the function can still change the side. 
3. Thirdly, this issue could be handled by changing the Moodle core. This seems most reasonable since Google encounters the same problem in their repository. 

- [ ] **Handle log out system account** This really distress me since for unknown reason the system account is sometimes unable to connect and is merely able to update the token in case the administrator logs in again. It seems to appear when shares are manually deleted. However, this has still to be proven. It might require additionally some research in the ownCloud code.
- [ ] **Research default storage** There might be a default location for ownCloud users to store their downloads. In this case the Moodlefolder should be embedded into this default location.

**Comment by NinaHerrmann** [comment]
For easy reference to error logging
reference_file_selected:
[reference_file_selected.pdf](https://github.com/learnweb/moodle-repository_owncloud/files/1562207/reference_file_selected.pdf)
send_file:
[send_file.pdf](https://github.com/learnweb/moodle-repository_owncloud/files/1562208/send_file.pdf)

**Comment by NinaHerrmann** [comment]
FYI: outsourcing parts of the access_controlled_link functions is still WIP.

**Comment by Dagefoerde** [comment]
warning to other developers: rebased onto current `master`.

**Comment by Dagefoerde** [comment]
Does not work. I tried to add an access controlled link and then submit the filepicker.
Screenshot (Moodle) after submitting the filepicker: 
clicking on save

ownCloud log output:
```
[Tue May 29 15:20:36 2018] 172.17.0.10:38464 [207]: /remote.php/webdav                                                                                                                                             
[Tue May 29 15:20:36 2018] 172.17.0.10:38466 [201]: /remote.php/webdav//System                                                                                                                                     
[Tue May 29 15:20:36 2018] 172.17.0.10:38468 [207]: /remote.php/webdav//System                                                                                                                                     
[Tue May 29 15:20:37 2018] 172.17.0.10:38470 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde                                                                                                         
[Tue May 29 15:20:37 2018] 172.17.0.10:38472 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde                                                                                                         
[Tue May 29 15:20:37 2018] 172.17.0.10:38474 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user                                                                                                    
[Tue May 29 15:20:37 2018] 172.17.0.10:38476 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user                                                                                                    
[Tue May 29 15:20:37 2018] 172.17.0.10:38478 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private                                                                                            
[Tue May 29 15:20:37 2018] 172.17.0.10:38480 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private                                                                                            
[Tue May 29 15:20:37 2018] 172.17.0.10:38482 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private/0                                                                                          
[Tue May 29 15:20:42 2018] Exception: {"Exception":"Error","Message":"Maximum function nesting level of '256' reached, aborting!","Code":0,"Trace":"#0 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(1149): Doctrine\\DBAL\\Query\\QueryBuilder->getSQLForJoins('`*PREFIX*extern...', Array)
#1 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(1112): Doctrine\\DBAL\\Query\\QueryBuilder->getFromClauses()
#2 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(244): Doctrine\\DBAL\\Query\\QueryBuilder->getSQLForSelect()
#3 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(206): Doctrine\\DBAL\\Query\\QueryBuilder->getSQL()
#4 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/DB\/QueryBuilder\/QueryBuilder.php(141): Doctrine\\DBAL\\Query\\QueryBuilder->execute()
#5 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/DBConfigService.php(337): OC\\DB\\QueryBuilder\\QueryBuilder->execute()
#6 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/DBConfigService.php(103): OC\\Files\\External\\Service\\DBConfigService->getMountsFromQuery(Object(OC\\DB\\QueryBuilder\\QueryBuilder))
#7 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(73): OC\\Files\\External\\Service\\DBConfigService->getAdminMounts()
#8 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(129): OC\\Files\\External\\Service\\StoragesService->readDBConfig()
#9 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(179): OC\\Files\\External\\Service\\StoragesService->readConfig()
#10 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(188): OC\\Files\\External\\Service\\StoragesService->getAllStorages()
#11 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/LegacyUtil.php(100): OC\\Files\\External\\Service\\StoragesService->getStorages()
#12 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Util.php(303): OC\\Files\\External\\LegacyUtil::getSystemMountPoints()
#13 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Keys\/Storage.php(343): OC\\Encryption\\Util->isSystemWideMountPoint('\/files\/System\/U...', 'admin')
#14 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Keys\/Storage.php(323): OC\\Encryption\\Keys\\Storage->getPathToKeys('\/admin\/\/files\/S...')
#15 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(1030): OC\\Encryption\\Keys\\Storage->copyKeys('\/admin\/\/files\/S...', '\/admin\/files\/Sy...')
#16 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(770): OC\\Files\\Storage\\Wrapper\\Encryption->copyKeys('\/admin\/\/files\/S...', '\/admin\/files\/Sy...')
#17 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#18 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#19 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#20 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#21 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#22 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#23 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#24 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#25 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#26 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#27 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#28 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#29 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#30 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#31 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#32 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
<snip>
#229 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#230 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#231 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#232 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#233 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#234 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#235 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#236 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#237 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#238 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#239 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System', 'files\/System\/Us...', false, false)
#240 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System', 'files\/System\/Us...', false, false)
#241 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files', 'files\/System\/Us...', false, false)
#242 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(358): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files', 'files\/System\/Us...')
#243 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php(283): OC\\Files\\Storage\\Wrapper\\Encryption->copy('files', 'files\/System\/Us...')
#244 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php(283): OC\\Files\\Storage\\Wrapper\\Wrapper->copy('files', 'files\/System\/Us...')
#245 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/View.php(945): OC\\Files\\Storage\\Wrapper\\Wrapper->copy('files', 'files\/System\/Us...')
#246 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/public\/Events\/EventEmitterTrait.php(50): OC\\Files\\View->OC\\Files\\{closure}(Array)
#247 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/View.php(988): OC\\Files\\View->emittingCall(Object(Closure), Array, 'file', 'copy')
#248 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/ObjectTree.php(250): OC\\Files\\View->copy('\/', '\/System\/User Ja...')
#249 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/CorePlugin.php(684): OCA\\DAV\\Connector\\Sabre\\ObjectTree->copy('', 'System\/User Jan...')
#250 [internal function]: Sabre\\DAV\\CorePlugin->httpCopy(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))
#251 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)
#252 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(479): Sabre\\Event\\EventEmitter->emit('method:COPY', Array)
#253 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))
#254 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(63): Sabre\\DAV\\Server->exec()
#255 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/remote.php(165): require_once('\/home\/j_dage01\/...')
#256 {main}","File":"\/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php","Line":1306}
[Tue May 29 15:20:42 2018] 172.17.0.10:38484 [500]: /remote.php/webdav/
```

File: classes/ocs_client.php [hunk 1]
```diff
@@ -41,56 +41,113 @@ class ocs_client extends rest {
...
+     * shareid at the suitable place for delete_share and get_information_of_share.
      * Docs:
      * create_share: https://doc.owncloud.org/server/10.0/developer_manual/core/ocs-share-api.html#create-a-new-share
      *
      */
     public function get_api_functions() {
         return [
             'create_share' => [
-                'endpoint' => $this->ocsendpoint,
+                'endpoint' => $this->ocsendpoint->out(),
```
Comment (Dagefoerde) [inline classes/ocs_client.php]: `->out(false)`

**Comment by Dagefoerde** [comment]
- [ ]  Files are shared with no restrictions regarding the usage; i.e. users who gain access to an access controlled link are allowed to re-share and to edit(!) the file. I have not compared this to Moodle+Google Docs et al, but I guess this is not intended behaviour.

**Comment by Dagefoerde** [comment]
- [x] Folder "Moodlefiles" displays a warning that it should never be changed. I don't think this is correct, because the behaviour is actually very resilient to changing foldernames: If a share already exists, you get the correct location via the OCS API, regardless of the used folder (users may even move around the file without issue). If it doesn't, the file is shared into the (then active) folder, which is perfect. Worst case scenario is that users accumulate a bunch of "Moodlefiles" folders because the admins change the name frequently... But that shouldn't be a reason for presenting this warning.
- [ ] On accessing a file, if the user is not logged in in ownCloud, a button is shown that leads the user to the login. In some cases, the user can't see this button, because a) it is in an invisible iframe, b) the file is referenced via `<img src=`. We cannot do much about b), ~~but we can circumvent a) by showing a popup right away (skipping the button).~~ and not much about a) either. --- Tried that, doesn't work: The popup (obviously) gets blocked by a popup blocker :( attempt is in `Dagefoerde/popup-on-sendfile`.

**Comment by Dagefoerde** [comment]
- [x] Access controlled shares never expire – Admins can set an expiration date which is (probably) transmitted to oC/Nc on creation of a share. However, both don't support expiration of private shares, so that setting is effectively useless. [ Update: ] The setting is not used at all and can be safely deleted.
- [x] Google and OneDrive repositories allow settings that control how files go into Moodle (option to select between only copying, access controlled links, or both)

**Comment by Dagefoerde** [comment]
- [x] Folder structure in system account heavily depends on users' chosen frontend languages, i.e. files might not be found because a wrong language is assumed (fix cf. https://github.com/moodle/moodle/commit/e0a903243be16b461cedfc223d59abe42bd527a2#diff-7066ab7ec56b9a7f65f10f907f4dc15a)

**Comment by Dagefoerde** [comment]
- [x] access controlled shares from the system account to (student) users may not be writeable by users!

**Comment by Dagefoerde** [comment]
Needs new docs, then it's ready for merging and for releasing a new version.

**Pull request merged**